test/periodic.jl

```diff
@@ -0,0 +1,332 @@
+@testset "periodic" begin
+    # Constructors
+    @test Periodic(sin, 5) == Periodic(sin, 5, 0)
+
+    @testset "_periodic" begin
+        @testset "$f" for f in (sin, cos)
+            # A typical 24-hour day
+            for h in 0:24
+                result = _periodic(
+                    f,
+                    ZonedDateTime(2015, 6, 1, tz"America/Winnipeg") + Hour(h),
+                    Day(1),
+                )
+                expected = f(2π / 24 * h)
+                @test result ≈ expected atol=1e-15
+            end
+
+            # A spring daylight-saving time change where the day has 23-hours
+            for h in 0:23
+                result = _periodic(
+                    f,
+                    ZonedDateTime(2015, 3, 8, tz"America/Winnipeg") + Hour(h),
+                    Day(1),
+                )
+                expected = f(2π / 23 * h)
+                @test result ≈ expected atol=1e-15
+            end
+
+            # A fall daylight-saving time change where the day has 25-hours
+            for h in 0:25
+                result = _periodic(
+                    f,
+                    ZonedDateTime(2015, 11, 1, tz"America/Winnipeg") + Hour(h),
+                    Day(1),
+                )
+                expected = f(2π / 25 * h)
+                @test result ≈ expected atol=1e-15
+            end
+        end
+
+        @testset "negative period" begin
+            @test ==(
+                _periodic(sin, DateTime(2000, 1, 1, 3), Day(-1)),
+                -_periodic(sin, DateTime(2000, 1, 1, 3), Day(1)),
+            )
+
+            @test ==(
+                _periodic(cos, DateTime(2000, 1, 1, 3), Day(-1)),
+                _periodic(cos, DateTime(2000, 1, 1, 3), Day(1)),
+            )
+        end
+
+        @testset "phase shift" begin
+            @test _periodic(sin, DateTime(1), Week(1)) == 0
+            @test _periodic(sin, DateTime(1), Week(1), Day(7)) == 0
+            @test _periodic(sin, DateTime(1), Week(1), Day(-7)) == 0
+
+            # Using the phase shift we can make a Thursday the beginning of the week.
+            # Note: that these calculations can change depending on the period type given.
```

Comment:
why is this? seems weird to me. `Second(Week(1))` is the same amount of time as `Week(1)`

Reply:
Ah, I was lazy - the whole `_periodic` test set is copied from the original `periodic`. So I'm confused too. I need to re-evaluate the old assumptions.

Reply:
It's because of `floor(dt:DateTime, p::Period)` in `_periodic`. It gives the nearest multiple of `p` in time.

`floor(DateTime(1), Week(1)) == 0001-01-01T00:00:00`

This is kinda weird because it's a Saturday - I thought it would give a Monday. But it's the year 0001 AD so maybe something weird happens to the calendar. It floors to Monday for recent dates e.g. 

```
floor(DateTime(2021), Week(1)) == 2020-12-28T00:00:00
floor(DateTime(2021, 2, 10), Week(1)) == 2021-02-08T00:00:00
```

Anyway, for seconds:

`floor(DateTime(1), Second(Week(1))) == 0000-12-30T00:00:00`

Because December 30 is exactly 52 weeks _in seconds_ from time 0.

Reply:
For now I've briefly explained this in the `_periodic` docstring.

Reply:
ah ok - makes then.